### PR TITLE
Feature/testing w3p

### PR DESCRIPTION
**Pull request by WhiteNik16**

### CODE REVIEW

File: packages/w3p/src/helpers/eth.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { CHAIN_TYPES } from '@/enums'
+import { getEthExplorerAddressUrl, getEthExplorerTxUrl } from '@/helpers'
+
+const exampleChain = {
+  id: 1,
+  name: 'Example',
+  rpcUrl: 'https://mainnet.example.io/v3/',
+  explorerUrl: 'https://example.io',
+  token: { name: 'Example', symbol: 'ex', decimals: 18 },
```
Comment (Sorizen) [inline packages/w3p/src/helpers/eth.test.ts]: use const for decimals

File: packages/w3p/src/provider-detector.test.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    test('should list empty providers', async () => {
+      mockWindow({ ethereum: {} })
+      expect(providerDetector.providers).toEqual({})
+    })
+    test('should add and return list providers', async () => {
+      mockWindow({ ethereum: { providers: [MetamaskProvider] } })
+      providerDetector.addProvider({ name: PROVIDERS.Metamask })
+      providerDetector.addProvider({ name: PROVIDERS.Coinbase })
+      expect(providerDetector.providers).toEqual({
+        metamask: { name: 'metamask' },
```
Comment (Sorizen) [inline packages/w3p/src/provider-detector.test.ts]: create const for provider names

File: packages/w3p/src/providers/raw/near-raw-provider.test.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        },
+        options: {} as Options,
+        setActiveAccount: jest.fn,
+        on: jest.fn(),
+        off: jest.fn(),
+      }
+      ;(
+        setupWalletSelector as jest.MockedFunction<typeof setupWalletSelector>
+      ).mockResolvedValue(mockSelector)
+      const result = await provider.init()
```
Comment (Sorizen) [inline packages/w3p/src/providers/raw/near-raw-provider.test.ts]: rename `result` variable and why `provider.init()` returning something?

File: packages/w3p/src/providers/wrapped/near.test.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+const mockProvider = {
+  signIn: jest.fn(),
+  signOut: jest.fn(),
+  init: jest.fn(),
+  signAndSendTxs: jest.fn(),
+  emit: jest.fn(),
+}
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const nearProvider = new NearProvider(mockProvider as any)
```
Comment (Sorizen) [inline packages/w3p/src/providers/wrapped/near.test.ts]: dont use `any` type

File: packages/w3p/src/tests/window.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const mockWindow = (value: any) => {
+  global.window = value
+}
+
+export const clearWindowMock = () => {
+  global.window = undefined as never
```
Comment (Sorizen) [inline packages/w3p/src/tests/window.ts]: why u use here `never` type?

File: packages/w3p/src/tests/window.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const mockWindow = (value: any) => {
```
Comment (Sorizen) [inline packages/w3p/src/tests/window.ts]: don't use `any` as type

File: packages/w3p/src/helpers/eth.test.ts (hunk 1 above)
Comment (WhiteNik16) [reply packages/w3p/src/helpers/eth.test.ts]: fixed

File: packages/w3p/src/provider-detector.test.ts (hunk 2 above)
Comment (WhiteNik16) [reply packages/w3p/src/provider-detector.test.ts]: fixed

File: packages/w3p/src/providers/raw/near-raw-provider.test.ts (hunk 3 above)
Comment (WhiteNik16) [reply packages/w3p/src/providers/raw/near-raw-provider.test.ts]: renamed
init returned boolean value, whether the initialization was successful

File: packages/w3p/src/providers/wrapped/near.test.ts (hunk 4 above)
Comment (WhiteNik16) [reply packages/w3p/src/providers/wrapped/near.test.ts]: fixed

File: coverage/clover.xml [hunk 7]
```diff
@@ -0,0 +1,427 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (napalmpapalam) [inline coverage/clover.xml]: doest it should be in the repo, can we add it to the gitignore

File: packages/w3p/src/helpers/amount.test.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { hexToDecimal } from '@/helpers'
+
+describe('performs amount unit test', () => {
+  test('hexToDecimal should return correct value', () => {
```
Comment (napalmpapalam) [inline packages/w3p/src/helpers/amount.test.ts]: no need to wrap to the test, describe is enough

File: packages/w3p/src/helpers/eth.test.ts (hunk 1 above)
Comment (napalmpapalam) [reply packages/w3p/src/helpers/eth.test.ts]: for what?

File: packages/w3p/src/helpers/eth.test.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        getEthExplorerAddressUrl(
+          exampleChain,
+          '0xFeebabE6b0418eC13b30aAdF129F5DcDd4f70CeA',
+        ),
+      ).toBe(
+        'https://example.io/address/0xFeebabE6b0418eC13b30aAdF129F5DcDd4f70CeA',
+      )
+    })
+  })
+})
```
Comment (napalmpapalam) [inline packages/w3p/src/helpers/eth.test.ts]: u a not testing an object,no need to nest test cases, split tests

File: packages/w3p/src/provider-detector.test.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    let providerDetector: ProviderDetector
+
+    beforeEach(() => {
+      providerDetector = new ProviderDetector()
+    })
+
+    afterEach(clearWindowMock)
+
+    test('should list empty providers', async () => {
+      mockWindow({ ethereum: {} })
```
Comment (napalmpapalam) [inline packages/w3p/src/provider-detector.test.ts]: u can mock it inside beforeEach and mock in some test if u need other mock value

File: packages/w3p/src/provider-detector.test.ts (hunk 2 above)
Comment (napalmpapalam) [reply packages/w3p/src/provider-detector.test.ts]: use enum PROVIDERS instead

File: packages/w3p/src/provider.test.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  afterEach(() => {
+    jest.restoreAllMocks()
+  })
+
+  describe('performs provider init', () => {
+    test('should throw an error if no provider instance is injected', async () => {
+      const provider = new Provider(MetamaskProvider)
+      await expect(
+        provider.init({ name: PROVIDERS.Metamask }),
+      ).rejects.toThrow()
```
Comment (napalmpapalam) [inline packages/w3p/src/provider.test.ts]: check throwed error or error message

File: packages/w3p/src/providers/wrapped/base-evm.test.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    expect(provider.isConnected).toBe(false)
+  })
+
+  test('should have a chainId property that returns undefined when not set', () => {
+    expect(provider.chainId).toBeUndefined()
+  })
+
+  test('should have an address property that returns undefined when not set', () => {
+    expect(provider.address).toBeUndefined()
+  })
```
Comment (napalmpapalam) [inline packages/w3p/src/providers/wrapped/base-evm.test.ts]: do it inside one test

File: packages/w3p/src/providers/wrapped/base-solana.test.ts [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+jest.mock('@/helpers', () => ({
+  getSolExplorerAddressUrl: jest.fn().mockReturnValue('mockAddressUrl'),
+  getSolExplorerTxUrl: jest.fn().mockReturnValue('mockTxUrl'),
+  handleEthError: jest.fn(),
+  handleSolError: jest.fn(),
+}))
+
+jest.mock('@/enums', () => ({
```
Comment (napalmpapalam) [inline packages/w3p/src/providers/wrapped/base-solana.test.ts]: why a u mocking enums?

File: packages/w3p/src/tests/window.ts (hunk 6 above)
Comment (napalmpapalam) [reply packages/w3p/src/tests/window.ts]: u can use unknown

File: packages/w3p/src/tests/window.ts [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+  global.window = value
+}
+
+export const clearWindowMock = () => {
+  global.window = undefined as never
+}
```
Comment (napalmpapalam) [inline packages/w3p/src/tests/window.ts:7]: changelog

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: packages/w3p/src/provider-detector.test.ts (hunk 2 above)
Comment (WhiteNik16) [reply packages/w3p/src/provider-detector.test.ts]: Yes, I used them.

File: packages/w3p/src/tests/window.ts (hunk 6 above)
Comment (WhiteNik16) [reply packages/w3p/src/tests/window.ts]: unknow does not fit the Window type

File: coverage/clover.xml (hunk 7 above)
Comment (WhiteNik16) [reply coverage/clover.xml]: fixed

File: packages/w3p/src/helpers/amount.test.ts (hunk 8 above)
Comment (WhiteNik16) [reply packages/w3p/src/helpers/amount.test.ts]: done

File: packages/w3p/src/helpers/eth.test.ts (hunk 9 above)
Comment (WhiteNik16) [reply packages/w3p/src/helpers/eth.test.ts]: fixed

File: packages/w3p/src/provider.test.ts (hunk 11 above)
Comment (WhiteNik16) [reply packages/w3p/src/provider.test.ts]: done

File: packages/w3p/src/providers/wrapped/base-evm.test.ts (hunk 12 above)
Comment (WhiteNik16) [reply packages/w3p/src/providers/wrapped/base-evm.test.ts]: fixed

File: packages/w3p/src/providers/wrapped/base-solana.test.ts (hunk 13 above)
Comment (WhiteNik16) [reply packages/w3p/src/providers/wrapped/base-solana.test.ts]: fixed

File: packages/w3p/src/providers/raw/near-raw-provider.test.ts (hunk 3 above)
Comment (Sorizen) [reply packages/w3p/src/providers/raw/near-raw-provider.test.ts]: i thing u should throw an error and handle it if provider not created, your init method should have `void` type i think

Comment (WhiteNik16) [reply packages/w3p/src/providers/raw/near-raw-provider.test.ts]: need to discuss

File: packages/w3p/src/tests/window.ts (hunk 5 above)
Comment (WhiteNik16) [reply packages/w3p/src/tests/window.ts]: need to discuss

File: packages/w3p/src/provider-detector.test.ts (hunk 10 above)
Comment (WhiteNik16) [reply packages/w3p/src/provider-detector.test.ts]: done

File: packages/w3p/src/tests/window.ts (hunk 14 above)
Comment (WhiteNik16) [reply packages/w3p/src/tests/window.ts:7]: done

**Pull request merged**